lib/services/access.go

```diff
@@ -59,3 +63,34 @@ type Access interface {
 	// ReplaceRemoteLocks replaces the set of locks associated with a remote cluster.
 	ReplaceRemoteLocks(ctx context.Context, clusterName string, locks []types.Lock) error
 }
+
+// CheckDynamicLabelsInDenyRules checks if any deny rules in the given role use
+// labels prefixed with "dynamic/".
+func CheckDynamicLabelsInDenyRules(r types.Role) error {
+	errorMessage := fmt.Sprintf("labels with %q prefix are not allowed in deny rules", types.TeleportDynamicLabelPrefix)
+	for _, labels := range []types.Labels{
+		r.GetNodeLabels(types.Deny),
+		r.GetAppLabels(types.Deny),
+		r.GetKubernetesLabels(types.Deny),
+		r.GetDatabaseLabels(types.Deny),
+		r.GetWindowsDesktopLabels(types.Deny),
+		r.GetGroupLabels(types.Deny),
+	} {
+		for label := range labels {
+			if strings.HasPrefix(label, types.TeleportDynamicLabelPrefix) {
+				return trace.BadParameter(errorMessage)
+			}
+		}
```

Comment:
Something like this to catch label expressions as well
```suggestion
	for _, kind := range []types.LabelMatcherKinds {
		labelMatchers, err := r.GetLabelMatchers(types.Deny, kind)
		for label := range labelMatchers.Labels {
			if strings.HasPrefix(label, types.TeleportDynamicLabelPrefix) {
				return trace.BadParameter(errorMessage)
			}
		}
		if strings.Contains(labelMatchers.Expression, `"` + types.TeleportDynamicLabelPrefix) {
				return trace.BadParameter(errorMessage)
		}
```